### PR TITLE
Sparse Sliding Window Attention

### PR DESCRIPTION
**Pull request by changlan**
Sliding Window Attention (SWA) using Splash Attention. Refactored the Splash Attention integration to support non-trivial mask functions other than causal mask.

Tested on v4-8 VM.

Benchmark:
```
# No SWA (Causal)

ref_fwd:0.0040s, flash_fwd:0.0026s
ref_bwd:0.0140s, flash_bwd:0.0060s

# Bias-based SWA (No sparsity optimization, current implementation)

ref_fwd:0.0038s, flash_fwd:0.0033s
ref_bwd:0.0141s, flash_bwd:0.0141s

# NumpyMask-based SWA (Splash)

ref_fwd:0.0039s, flash_fwd:0.0022s
ref_bwd:0.0138s, flash_bwd:0.0051s

# Now with the sparse kernel, SWA is indeed faster!
```

### CODE REVIEW

File: axlearn/common/flash_attention/tpu_attention.py [hunk 1]
```diff
@@ -269,24 +274,11 @@ def _tpu_splash_attention(
...
-                return mask(*args, **kwargs)
-
-            return wrapped_mask
-
-        mask = ComputableMask(mask_shape, mask_function=wrap_mask(mask))
+        rows = np.arange(source_len)
+        cols = np.arange(target_len)
+        with jax.ensure_compile_time_eval():
+            mask_array = np.asarray(mask(rows[:, None], cols[None, :]))
+        mask = splash_attention_mask.NumpyMask(array=mask_array)
```
Comment (ruomingp) [inline axlearn/common/flash_attention/tpu_attention.py:281]: Comment on the maximum size of this NumpyMask? E.g., it will contain 1B entries for seq_len=32K?

**Review by ruomingp — APPROVED** [review]

File: axlearn/common/flash_attention/tpu_attention.py (hunk 1 above)
Comment (changlan) [reply axlearn/common/flash_attention/tpu_attention.py:284]: Yes indeed, it could consume a large amount of host memory at compile time. Added to the comment. I also added a test to compile with long sequences

**Review by ruomingp — APPROVED** [review]

**Review by ruomingp — APPROVED** [review]
